File: src/main/java/com/zunza/buythedip/auth/oauth2/CustomOAuth2SuccessHandler.java
package com.zunza.buythedip.auth.oauth2;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.Charset;

import org.springframework.security.core.Authentication;
import org.springframework.security.web.authentication.SimpleUrlAuthenticationSuccessHandler;
import org.springframework.stereotype.Component;

import com.zunza.buythedip.auth.jwt.JwtProvider;
import com.zunza.buythedip.infrastructure.redis.service.RedisCacheService;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class CustomOAuth2SuccessHandler extends SimpleUrlAuthenticationSuccessHandler {
	private final JwtProvider jwtProvider;
	private final RedisCacheService redisCacheService;

	@Override
	public void onAuthenticationSuccess(
		HttpServletRequest request,
		HttpServletResponse response,
		Authentication authentication
	) throws IOException, ServletException {
		CustomOAuth2User oAuth2User = (CustomOAuth2User)authentication.getPrincipal();
		Long userId = oAuth2User.getUserId();
		String nickname = oAuth2User.getName();

		String accessToken = jwtProvider.generateAccessToken(userId, oAuth2User.getAuthorities());
		String refreshToken = jwtProvider.generateRefreshToken(userId);
		redisCacheService.set(userId.toString(), refreshToken);

		String encodedNickname = URLEncoder.encode(nickname, Charset.defaultCharset());
		String redirectionUrl = "http://localhost:5173/login?token=" + accessToken + "&nickname=" + encodedNickname;

		response.sendRedirect(redirectionUrl);
	}
}

File: src/main/kotlin/com/zunza/buythedip/auth/oauth2/CustomOAuth2SuccessHandler.kt
package com.zunza.buythedip.auth.oauth2

import com.zunza.buythedip.auth.jwt.JwtTokenProvider
import com.zunza.buythedip.user.repository.RefreshJwtRepository
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.SimpleUrlAuthenticationSuccessHandler
import org.springframework.stereotype.Component
import java.net.URLEncoder
import java.nio.charset.Charset

@Component
class CustomOAuth2SuccessHandler(
    private val jwtTokenProvider: JwtTokenProvider,
    private val refreshJwtRepository: RefreshJwtRepository
) : SimpleUrlAuthenticationSuccessHandler() {
    override fun onAuthenticationSuccess(
        request: HttpServletRequest?,
        response: HttpServletResponse?,
        authentication: Authentication?
    ) {
        val oAuth2User = authentication?.principal as CustomOAuth2User

        val userId = oAuth2User.id
        val nickname = oAuth2User.name

        val accessJwt = jwtTokenProvider.generateAccessJwt(userId, oAuth2User.authorities)
        val refreshJwt = jwtTokenProvider.generateRefreshJwt(userId)
        refreshJwtRepository.save(userId, refreshJwt)

        val encodedNickname = URLEncoder.encode(nickname, Charset.defaultCharset())
        val redirectionUrl = "http://localhost:5173/login?token=$accessJwt&nickname=$encodedNickname"

        response?.sendRedirect(redirectionUrl)
    }
}
